### ISSUE TITLE
How to connect to multiple s3 buckets?

### ISSUE THREAD

**Issue by cookiejest** [duckdb/duckdb#4337]
hi there,

I love duckdb its awesome!!!

Is there a way to add multiple S3 access keys in the httpfs extension to support pulling data from multiple buckets/s3 environments?

Thanks!

**Comment by cookiejest** [comment]
This works great but is it possible to pass them as part of the read_csv_auto() function or some other way?
SET s3_region='eu-west-1';
SET s3_access_key_id=xxx;
SET s3_secret_access_key=xxx;

**Comment by samansmink** [comment]
Hi @cookiejest, we currently don't support having different sets of credentials at the same time for different files. To fetch data with different credentials you will need to switch the credentials between the queries:
```
SET s3_access_key_id=first_env_key; SET s3_secret_access_key=xxx;
SELECT * FROM 's3://bucket1/file1.csv'
SET s3_access_key_id=second_env_key; SET s3_secret_access_key=xxx;
SELECT * FROM 's3://bucket2/file2.csv'
```

you can also copy (parts of) a remote table after which you can query it without credentials
```
SET s3_access_key_id=first_env_key; SET s3_secret_access_key=xxx;
CREATE TABLE file1_copy AS SELECT * FROM 's3://bucket1/file1.csv where col='somevalue'
```

**Comment by cookiejest** [comment]
ok i thought about that, im new to duckdb, what is the performance of this like? Does it essentially copy the whole table into memory by doing a CREATE TABLE command or can it still work on tables bigger than RAM memory?

Thanks!

**Comment by samansmink** [comment]
@cookiejest yes this does do a copy. You can start duckdb in two ways: as a transient in-memory database or as a persistent database by passing a path: `duckdb your_db_file.db`.  The persistent db can be bigger than RAM. For more info on this check out our docs: https://duckdb.org/docs/

**Comment by mskyttner** [comment]
Running into a similar situation and wondering if this issue could be turned into a feature request for some kind of mechanism to make this a bit more convenient. 

For example, what about allowing a "compound setting" for an S3_URI following a connectionstring like format (similar to what Arrow supports, ie a string looking like s3://[access_key:secret_key@]bucket/path[?region=]). 

This in combination with a duckdb function to allow reading values for environment variables ("getenv" utility function) might be one mechanism to facilitate reading from multiple S3 sources?

**Comment by Alex-Monahan** [comment]
One of the design goals of DuckDB is to not depend on environment variables, so we would need an alternative method!

**Comment by cookiejest** [comment]
can it be done using a forloop like in postgresql?

BEGIN
    FOR r IN
        SELECT * FROM foo WHERE fooid > 0
    LOOP
        -- can do some processing here
        RETURN NEXT r; -- return current row of SELECT
    END LOOP;
    RETURN;
END;

**Comment by mskyttner** [comment]
@Alex-Monahan I'm not suggesting to introduce a dependency on environment variables, but feel that it would be great if there was a function that would simplify using data stored in an environment variable. There are functions to read data from csv, json, stdin so why not allow reading from a system environment variable? 

There is the "dot command" .shell that can be used already (both in duckdb and sqlite), for example

```sql
.shell echo $HOME
```
This displays the value of the $HOME environment variable. But there is no way, I think, to pick up and use such a value in a query? Or maybe there is already?
